libs/ml/include/ml/core/node.hpp

```diff
@@ -141,18 +154,184 @@ class Node
+      {
+        // TODO(VH): throw error: shaped layer returned empty shape, linking failed!
+        FETCH_LOG_ERROR(
+            this->name_.c_str(),
+            "A pre-shaped Node's underlying Op returned empty shape, shape linking impossible!");
+        return slice_output_shape_;
+      }
+      slice_output_shape_ = candidate;
+    }
+
+    ShapeVector input_shapes;
+    for (auto const &i : input_nodes_)
+    {
+      auto node_ptr = i.lock();
+      if (!node_ptr)
+      {
+        throw std::runtime_error("Unable to lock weak pointer.");
+      }
+      // Deeper recursive call.
+      auto const in_shape = node_ptr->SliceOutputShape();
+      if (!in_shape.empty())
+      {
+        input_shapes.emplace_back(in_shape);
+      }
+      // TODO(VH): else (if there _is_ an empty shape among inputs) what?
+    }
+
+    if (!input_shapes.empty())
+    {
+      if (slice_output_shape_.empty())
+      {
+        slice_output_shape_ = op_ptr_->ComputeSliceOutputShape(input_shapes);
+      }
+      else
+      {
+        // TODO(VH): implement input shape matching validation.
+      }
+    }
+    else
+    {
+      if (input_nodes_.size() != 1 || !i_am_preshaped)
+      {
+        FETCH_LOG_INFO(name_.c_str(), "Shape deduction reached a Graph leaf : " + this->name_);
+        return slice_output_shape_;
+      }
+      // If I am a pre-shaped node, and my Input is a lone Placeholder, I can force its shape.
+      std::shared_ptr<Node<TensorType>> input_node = input_nodes_.back().lock();
+      if (!input_node)
+      {
+        throw std::runtime_error("Unable to lock weak pointer.");
+      }
+      if (input_node->OpCode() == OpType::OP_PLACEHOLDER)
+      {
+        input_node->SetSliceOutputShape(op_ptr_->ExpectedSliceInputShapes().front());
+        FETCH_LOG_INFO(name_.c_str(), "Forced next layer shape to my expected input shape.");
+      }
+    }
+
+    if (slice_output_shape_.empty())
+    {
+      // throw an error: invalid calcs from a previous layer.
+      FETCH_LOG_INFO(name_.c_str(), "Error: shape deduction failed.");
+    }
+    return slice_output_shape_;
+  }
+
 private:
   std::vector<NodeWeakPtrType> input_nodes_;
   std::vector<NodeWeakPtrType> outputs_;
 
   std::string       name_;
   TensorType        cached_output_;
   CachedOutputState cached_output_status_;
-  OpType            operation_type_;
+  OpType operation_type_;  // TODO(VH): who synchronizes this code with underlying Ops OpCode?...
+
+  // TODO(VH): only Ops should have computed Shapes, and the Node
+  // must only ask its Op for output shape. "InformationExpert" violation.
+  Shape slice_output_shape_{};
 
   std::shared_ptr<ops::Ops<TensorType>> op_ptr_;
+
+  static const std::set<OpType> shaped_operations_;
 };
 
+/// A list of Ops that "knows" their output shape on construction, such as FullyConnected.
+template <typename TensorType>
+const std::set<OpType> Node<TensorType>::shaped_operations_{OpType::LAYER_FULLY_CONNECTED};
+
```

Comment:
quite possible i've missed something important, but it feels like we should just set up an input layer, and assume that there are no other ops that 'know' their output shape on construction. In this way we can try to cleanly separate what happens during graph.AddNode and graph.Compile (i.e. at AddNode time the shapes will always be unknown until we call Compile)

Reply:
If we do this, we don't need to include set and can ditch the shaped_operations_ set, i think

Reply:
Well, in genera - yes, but then it require some rework of FullyCOnnected layer initialisation. Right now weight tensor (its shape depends on input count, and basically this tensor is the reason why Dense is fixed-shape) is inited in constructor of FullyConnected layer, therefore we need to pass input count as a parameter to constructor.